File: rag-projects/new-rag-2026/backend/app/models/eval.py
"""
Pydantic models for the Ragas Evaluation Report feature.
"""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


# ── Turn (from SQLite) ────────────────────────────────────────────────────────

class EvalTurn(BaseModel):
    turn_id: str
    session_id: str
    turn_number: int
    question: str
    answer: str
    sources: list[dict[str, Any]] = []
    domain_keys: list[str] = []
    query_type: str = ""
    is_fallback: bool = False
    timestamp: str


# ── Session summary ───────────────────────────────────────────────────────────

class SessionSummary(BaseModel):
    session_id: str
    turn_count: int
    last_active: str
    created_at: str


class SessionListResponse(BaseModel):
    sessions: list[SessionSummary]
    total: int


class SessionTurnsResponse(BaseModel):
    session_id: str
    turns: list[EvalTurn]


# ── Eval run request ──────────────────────────────────────────────────────────

class EvalRunRequest(BaseModel):
    turn_ids: list[str] = Field(..., min_length=1)
    judge_provider: str = Field("openai", description="openai | gemini | siliconflow")
    judge_model: str = Field("gpt-4o-mini")
    reference_answers: dict[str, str] = Field(
        default_factory=dict,
        description="Optional {turn_id: reference_text} for context_recall / answer_correctness",
    )


class EvalRunStartResponse(BaseModel):
    run_id: str
    status: str = "pending"
    turn_count: int


# ── Per-turn score ────────────────────────────────────────────────────────────

class EvalTurnScore(BaseModel):
    turn_id: str
    question: str
    answer: str
    domain_keys: list[str] = []
    query_type: str = ""
    faithfulness: float | None = None
    answer_relevancy: float | None = None
    context_precision: float | None = None
    context_recall: float | None = None
    answer_correctness: float | None = None
    reference_answer: str = ""


# ── Run result ────────────────────────────────────────────────────────────────

class EvalRunResult(BaseModel):
    run_id: str
    status: str
    judge_provider: str
    judge_model: str
    created_at: str
    turn_count: int
    error: str | None = None
    scores: list[EvalTurnScore] = []
    averages: dict[str, float | None] = {}


class EvalRunSummary(BaseModel):
    run_id: str
    created_at: str
    turn_count: int
    judge_provider: str
    judge_model: str
    status: str
    error: str | None = None


class EvalRunListResponse(BaseModel):
    runs: list[EvalRunSummary]

File: rag-projects/new-rag-2026/frontend/src/types/eval.ts
// TypeScript types matching the eval API response shapes

export interface EvalTurn {
  turn_id: string
  session_id: string
  turn_number: number
  question: string
  answer: string
  sources: Record<string, unknown>[]
  domain_keys: string[]
  query_type: string
  is_fallback: boolean
  timestamp: string
}

export interface SessionSummary {
  session_id: string
  turn_count: number
  last_active: string
  created_at: string
}

export interface SessionListResponse {
  sessions: SessionSummary[]
  total: number
}

export interface SessionTurnsResponse {
  session_id: string
  turns: EvalTurn[]
}

export interface EvalRunRequest {
  turn_ids: string[]
  judge_provider: string
  judge_model: string
  reference_answers: Record<string, string>
}

export interface EvalRunStartResponse {
  run_id: string
  status: string
  turn_count: number
}

export interface EvalTurnScore {
  turn_id: string
  question: string
  answer: string
  domain_keys: string[]
  query_type: string
  faithfulness: number | null
  answer_relevancy: number | null
  context_precision: number | null
  context_recall: number | null
  answer_correctness: number | null
  reference_answer: string
}

export interface EvalRunResult {
  run_id: string
  status: string
  judge_provider: string
  judge_model: string
  created_at: string
  turn_count: number
  error: string | null
  scores: EvalTurnScore[]
  averages: Record<string, number | null>
}

export interface EvalRunSummary {
  run_id: string
  created_at: string
  turn_count: number
  judge_provider: string
  judge_model: string
  status: 'pending' | 'running' | 'done' | 'failed'
  error: string | null
}

export interface EvalRunListResponse {
  runs: EvalRunSummary[]
}

export const METRIC_NAMES = [
  'faithfulness',
  'answer_relevancy',
  'context_precision',
  'context_recall',
  'answer_correctness',
] as const

export type MetricName = (typeof METRIC_NAMES)[number]

export const METRIC_LABELS: Record<MetricName, string> = {
  faithfulness: 'Độ trung thực',
  answer_relevancy: 'Liên quan câu trả lời',
  context_precision: 'Độ chính xác ngữ cảnh',
  context_recall: 'Độ bao phủ ngữ cảnh',
  answer_correctness: 'Độ chính xác câu trả lời',
}
